### ISSUE TITLE
[BUG] Database migration don't run when executing spring-boot jar

### ISSUE THREAD

**Issue by wijnandtop** [jobrunr/jobrunr#83]
**Describe the bug**
When executing my project using "mvn spring-boot:run" everything runs fine, database migrations are executed on my postgress db. During the test phase"mvn clean package" migrations are executed on my in memory H2 database.
But when I run my application "java -jar target/app.jar" migrations aren't executed and Jobrunr fails with exception;

```
2021-01-11 18:01:02.355 ERROR 32675 --- [pool-1-thread-1] org.jobrunr.server.ServerZooKeeper       : An unrecoverable error occurred. Shutting server down...

org.jobrunr.storage.StorageException: org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "JOBRUNR_BACKGROUNDJOBSERVERS" not found; SQL statement:
```

This is the case with both H2 and Postgres, the in-memory storage provider works fine.
There are no log entries regarding a failed or skipped migration.

**Environment**
I'm using version: 1.2.3
I'm running on JRE / JDK: openjdk 11.0.9.1 2020-11-04 (with maven and java scope)
I'm using the following StorageProvider: SQL

**To Reproduce**
The error occurs both while defining the config in "application.properties" and using a config Bean (the first is also there in my example but commented).
I've isolated the problem in a "plain" spring-boot project, without any additional code:
https://github.com/wijnandtop/jobrunr_issue

```
mvn clean package
java -jar target/demo-0.0.1-SNAPSHOT.jar
```

**Expected behavior**
JobRunr check if migrations are needed if sol the migrations are executed and JobRunr starts.
I don't want to run them manually since I create dynamic environments for CI.

**Additional context**
Maybe I'm missing something and I made a mistake, I did spend a couple of hours on this, so if it is the case it may be worth documenting it.

**Comment by cathalbahn** [comment]
I am having the same issue. Figured out what is happening. The following section of code doesn't support nested jar files.
https://github.com/jobrunr/jobrunr/blob/master/core/src/main/java/org/jobrunr/utils/ClassPathUtils.java#L81

Here is an example of code on how to support an inner and outer jar files: https://stackoverflow.com/a/30015095

When the toPath is called with this url: "jar:file:/C:/PathToJar/outer.jar!/BOOT-INF/lib/jobrunr-1.2.3.jar!/org/jobrunr/storage/sql/common/migrations"
The `indexOf('!')` code will take "jar:file:/C:/PathToJar/outer.jar" as the jar and this "/BOOT-INF/lib/jobrunr-1.2.3.jar!/org/jobrunr/storage/sql/common/migrations" as the path.

The error is ignored on line 105: https://github.com/jobrunr/jobrunr/blob/master/core/src/main/java/org/jobrunr/utils/ClassPathUtils.java#L105

**Comment by rdehuyss** [comment]
A fix is being build as we speak...

@wijnandtop: thanks for the repo that reproduces it (it saved me a lot of time) to reproduce/debug it!
@cathalbahn: thanks for the extra info on Stackoverflow, it also really helped to create a fix

**Comment by rdehuyss** [comment]
Solved in release v1.2.4.

**Comment by cathalbahn** [comment]
The fix is throwing an exception. 

java.nio.file.ProviderNotFoundException: Provider not found

Failing here: https://github.com/jobrunr/jobrunr/blob/master/core/src/main/java/org/jobrunr/utils/JarFileSystemUtils.java#L46

**Comment by cathalbahn** [comment]
> 
> 
> Hi @cathalbahn - thx for the info but I need more info (or ideally a reproducible code example) to research this...
> 
>     * You're using Postgres?
> 
>     * is it not working when using a fat jar (spring-boot) jar or during development?
> 
>     * how are you running JobRunr (spring-boot) / tomcat / ... ?

* Using Postgres.

* When I run java -jar app.jar it fails. 

* When I run from IntelliJ it works.

* Using the spring-boot-starter. 

Used this guide to debug the jar: https://www.linkedin.com/pulse/debug-jar-files-intellij-idea-maksym-lushpenko

**Comment by rdehuyss** [comment]
Hmmm, I really cannot reproduce it anymore - see https://github.com/rdehuyss/jobrunr_issue - it's a fork from the original repo which had the problem.

Only thing I can think of now is difference between Linux and Windows :frowning: ? I'll try to setup a Windows machine in a VM...

**Comment by rdehuyss** [comment]
> The getFileSystem method is called with fileSystemName = jar:file:/C:/pathToJar/app.jar!/BOOT-INF/lib/jobrunr-1.2.4.jar

This looks completely correct...

**Comment by rdehuyss** [comment]
Which Java version are you using?

**Comment by cathalbahn** [comment]
> Which Java version are you using?

Java Version: 1.8
OS: Windows

**Comment by rdehuyss** [comment]
Managed to reproduce it - I was running with Java 15. Looking into a solution...

**Comment by rdehuyss** [comment]
Ok, so I'm afraid I don't have good news.

I debugged through the JVM (com.sun.nio.zipfs.ZipFileSystemProvider) and it's not supported in Java 8.

See also:
- https://stackoverflow.com/questions/58976001/find-filenames-in-a-folder-inside-nested-jars
- last comment on the stackoverflow you mentioned: https://stackoverflow.com/questions/29708142/how-can-i-create-a-filesystem-to-a-jar-inside-another-jar/30015095#comment-69924702

Not sure how to proceed for the moment...

**Comment by rdehuyss** [comment]
Since I want JobRunr to support Java 8, I found another way to run the migrations.

@cathalbahn: before I create an actual release for this, could you build JobRunr yourself and test whether it works for you?

**Comment by cathalbahn** [comment]
> 
> 
> Since I want JobRunr to support Java 8, I found another way to run the migrations.
> 
> @cathalbahn: before I create an actual release for this, could you build JobRunr yourself and test whether it works for you?

I build master and ran it locally. It is failing here: `jarFile.stream()` https://github.com/jobrunr/jobrunr/blob/master/core/src/main/java/org/jobrunr/storage/sql/common/migrations/RunningOnJava11OrLowerWithinFatJarMigrationProvider.java#L30

With: Method threw 'java.lang.IllegalStateException' exception. "zip file closed"

**Comment by cathalbahn** [comment]
Updating the spring boot version fixes the issue above. 
Found the solution here: https://github.com/ronmamo/reflections/issues/279#issuecomment-739926905

**Comment by rdehuyss** [comment]
What is the other exception? I tried with the code from the original bug (https://github.com/rdehuyss/jobrunr_issue) on Java 8 and all appears to work...

**Comment by cathalbahn** [comment]
> 
> 
> What is the other exception? I tried with the code from the original bug (https://github.com/rdehuyss/jobrunr_issue) on Java 8 and all appears to work...

Yea, everything is working now. Update above.

**Comment by rdehuyss** [comment]
@cathalbahn Thanks for the help in finding the root cause!

**Comment by rdehuyss** [comment]
Resolved with release v1.2.5, which will be available in Maven Central later today.
